torchmultimodal/modules/encoders/mdetr_text_encoder.py

```diff
@@ -0,0 +1,223 @@
+    def __init__(
+        self,
+        embedding_dim: int,
+        num_encoder_layers: int,
+        num_attention_heads: int,
+        ffn_dimension: Optional[int] = None,
+        dropout: float = 0.1,
+        normalize_before: bool = False,
+        return_all_layers: bool = False,
+    ):
+        super().__init__()
+        ffn_dimension = ffn_dimension or 4 * embedding_dim
+        layer = torch.nn.TransformerEncoderLayer(
+            d_model=embedding_dim,
+            nhead=num_attention_heads,
+            dim_feedforward=ffn_dimension,
+            dropout=dropout,
+            activation="gelu",
+            batch_first=True,
+            norm_first=normalize_before,
+        )
+        self.layers = torch.nn.TransformerEncoder(
+            encoder_layer=layer, num_layers=num_encoder_layers
+        )
+        self.normalize_before = normalize_before
+        self.embedding_layer_norm = nn.LayerNorm(embedding_dim)
+        self.return_all_layers = return_all_layers
+
+    def _forward_return_all_layers(
+        self,
+        embeddings: torch.Tensor,
+        attn_mask: Optional[torch.Tensor] = None,
+        padding_mask: Optional[torch.Tensor] = None,
+    ) -> List[torch.Tensor]:
+        encoded = embeddings
+        states = [encoded]
+        for layer in self.layers.layers:
+            encoded = layer(encoded, src_key_padding_mask=attn_mask)
+            states.append(encoded)
+        if self.normalize_before:
+            for i, state in enumerate(states):
+                states[i] = self.embedding_layer_norm(state)
+        return states
+
+    def _forward_return_last_layer(
+        self,
+        embeddings: torch.Tensor,
+        attn_mask: Optional[torch.Tensor] = None,
+        padding_mask: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        encoded = embeddings
+        for layer in self.layers.layers:
+            encoded = layer(encoded, src_key_padding_mask=attn_mask)
+        if self.normalize_before:
+            encoded = self.embedding_layer_norm(encoded)
+        return encoded
+
+    def forward(
+        self,
+        embeddings: torch.Tensor,
+        attn_mask: Optional[torch.Tensor] = None,
+        padding_mask: Optional[torch.Tensor] = None,
+    ) -> Union[torch.Tensor, List[torch.Tensor]]:
+        out: Union[torch.Tensor, List[torch.Tensor]]
+        if self.return_all_layers:
+            out = self._forward_return_all_layers(embeddings, attn_mask, padding_mask)
+        else:
+            out = self._forward_return_last_layer(embeddings, attn_mask, padding_mask)
+
+        return out
+
+
+class MDETRTextEncoder(nn.Module):
+    def __init__(self, embeddings: nn.Module, encoder: nn.Module):
+        super().__init__()
+        self.embeddings = embeddings
+        self.encoder = encoder
+
+    # Note: MDETR's RoBERTa encoder also returns pooler outputs in forward, but
+    # these are only used for contrastive loss. Since contrastive loss is not used anywhere,
```

Comment:
what do u mean by contrastive loss not being used anywhere?

Reply:
So there are two contrastive losses: contrastive alignment loss and contrastive loss. Contrastive alignment loss is used but as far as I can tell contrastive loss is not. So I am referring to the latter here